File: semester-2/lab-5/task-1/C#/Shape.cs
using System;

namespace lab_5
{
    public abstract class Shape
    {
        protected Point[] Points;
    
        protected Shape(params int[] coords)
        {
            if (coords.Length % 2 == 0)
            {
                Points = new Point[coords.Length / 2];
                int p = 0;
                for (int i = 0; i < coords.Length; i += 2)
                {
                    Points[p] = new Point(coords[i], coords[i + 1]);
                    p++;
                }
            }
            else
            {
                throw new Exception("Coords amount should be even.");
            }
        }

        protected readonly struct Point
        {
            public Point(double x, double y)
            {
                X = x;
                Y = y;
            }

            public double X { get; }
            public double Y { get; }
        }

        private Point[] this[int index]
        {
            get
            {
                if (index <= Points.Length - 1 && index >= 0)
                {
                    return index != Points.Length - 1
                        ? new Point[] {Points[index], Points[index + 1]}
                        : new Point[] {Points[0], Points[^1]};
                }
                throw new Exception("There are no side #" + index + " in this shape!");
            }
        }

        protected float SideLength(int sideIndex)
        {
            Point[] side = this[sideIndex];
            return (float) Math.Sqrt(Math.Pow(side[0].X - side[1].X, 2) + Math.Pow(side[0].Y - side[1].Y, 2));
        }
    }
}

File: semester-2/lab-5/task-1/C++/Shape.cpp
//
// Created by Arthur Ladchenko on 14.04.2020.
//

#include "Shape.h"
#include <cmath>

Shape::Shape(int *coords, int size) {
    if (size % 2 == 0) {
        points = new Point[size / 2];
        int p = 0;
        for (int i = 0; i < size; i += 2) {
            Point np;
            np.x = coords[i];
            np.y = coords[i + 1];
            points[p] = np;
            p++;
        }
        shapePoints = p;
    } else {
        throw std::runtime_error("Coords amount should be even.");
    }
}

float Shape::SideLength(int sideIndex) {
    Point* side = GetSide(sideIndex);
    return sqrt(pow(side[0].x - side[1].x, 2) + pow(side[0].y - side[1].y, 2));
}

Point *Shape::GetSide(int sideIndex) {
    if (sideIndex <= GetSize() - 1 && sideIndex >= 0) {
        if (sideIndex != GetSize() - 1) {
            Point* side = new Point[2];
            side[0] = points[sideIndex];
            side[1] = points[sideIndex + 1];
            return side;
        } else {
            Point* side = new Point[2];
            side[0] = points[0];
            side[1] = points[GetSize() - 1];
            return side;
        }
    }
    throw std::runtime_error("There are no such side in this shape!");
}

int Shape::GetSize() {
    return this->shapePoints;
}
